### PR TITLE
Implement `compareResultsUsingQuery` helper function

### PR DESCRIPTION
**Pull request by benjamn**
This helper function demonstrates how we might use query syntax (and specifically a hypothetical `@nonreactive` directive) to direct the comparison of JSON result objects, ignoring certain (`@nonreactive`) parts of the objects, without modifying either object in the process.

### CODE REVIEW

File: src/react/hooks/__tests__/compareResults.test.tsx [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+      { things: {} },
+      { things: [] },
+    )).toBe(false);
+
+    expect(compareResultsUsingQuery(
+      query,
+      { things: [] },
+      { things: {} },
+    )).toBe(false);
+
```
Comment (alessbell) [inline src/react/hooks/__tests__/compareResults.test.tsx:317]: Thoughts on adding a test that shows two empty arrays returns `true`?

Wouldn't be testing any additional behavior, of course, just an example of a common case.

**Review by alessbell — APPROVED** [review]
Can't wait to use this! 🙇🚢

File: src/react/hooks/__tests__/compareResults.test.tsx (hunk 1 above)
Comment (benjamn) [reply src/react/hooks/__tests__/compareResults.test.tsx:317]: 31237810ebe02ccbae2ee35a90bd09e23fb00f23
